Improve song select load and reload with large beatmap databases

peppy opened the pull request:
Supersedes and closes https://github.com/ppy/osu/pull/29550.

Test video with 100k+ beatmaps (can provide database on request).

https://github.com/user-attachments/assets/1491282f-944f-4d44-8878-f34f9a088c21


- Detach is now asynchronous and global (removes ~1 s stutter / load time for song select)
 - As a result, we **no longer need to preload song select**. This has large benefits, including reduced memory pressure when not using song select (especially noticeable when exiting the game from song select – previously it would load another one while trying to exit the game, causing a delay in shut down by 1-5 seconds).
 - This also removes stutters seen when loading song select from places that aren't the main menu, resolving the performance part of #21952.
 - This also means that we are no longer requiring to query realm twice (`BeatmapCarousel` used to do an initial query in `load` followed by a second implicit one from the initial subscription callback; we now only use the callback).
- Initial filter operation is now asynchronous (removes ~500 ms stutter on carousel display)

Reviewer notes:

I've made quite a few (unavoidable) changes to `BeatmapCarousel` and `SongSelect`, but I stand behind them being steps-in-the-right-direction from what we had. Individual commits should hopefully make things easy to digest.

Commit dd4a1104e45ddd50015608555227fe17afdb6754 may be a bit hard to accept. If so, this commit can be removed. But it does reduce complexity if accepted.

peppy committed:
Add basic detached beatmap store

peppy committed:
Use bindable list implementation

peppy committed:
Fix incorrect cancellation / disposal handling of `DetachedBeatmapStore`

peppy committed:
Fix null reference in change handling code

peppy committed:
Fix enumeration over modified collection

peppy committed:
Detach beatmap sets asynchronously

peppy committed:
Remove song select preloading

Really unnecessary now.

peppy committed:
Allow running initial filter criteria asynchronously

This reverts a portion of https://github.com/ppy/osu/pull/9539.

The rearrangement in `SongSelect` is required to get the initial filter
into `BeatmapCarousel` (and avoid the `FilterChanged` event firing,
causing a delayed/scheduled filter application).

peppy committed:
Always debounce external `Filter` requests (except for tests)

The only exception to the rule here was "when screen isn't active apply
without debounce" but I'm not sure we want this. It would cause a
stutter on returning to song select and I'm not even sure this is a
common scenario.

I'd rather remove it and see if someone finds an actual case where this
is an issue.

peppy committed:
Fix multiple test failures

peppy committed:
Tidy up test beatmap loading

peppy committed:
Remove `loadedTestBeatmaps` flag

peppy committed:
Reduce test filter count expectation by one due to initial filter being implicit

peppy committed:
Fix some tests not always waiting long enough for beatmap loading

These used to work because there was a huge blocking load operation,
which is now more asynchronous.

Note that the change made in `SongSelect` is not required, but defensive
(feels it should have been doing this the whole time).

peppy committed:
Fix realm changes being applied before detach finishes

peppy committed:
Merge branch 'master' into detached-beatmap-cache

bdach commented on osu.Game/Database/DetachedBeatmapStore.cs:
```diff
@@ -0,0 +1,154 @@
+                // Detaching beatmaps takes some time, so let's make sure it doesn't run on the update thread.
+                var frozenSets = sender.Freeze();
+
+                Task.Factory.StartNew(() =>
+                {
+                    realm.Run(_ =>
+                    {
+                        var detached = frozenSets.Detach();
+
+                        detachedBeatmapSets.Clear();
+                        detachedBeatmapSets.AddRange(detached);
+                        loaded.Set();
+                    });
+                }, TaskCreationOptions.LongRunning);
```
No error handling here whatsoever worries me that if if this ever fails (for freak accident reasons) it will deadlock the game because `loaded` will never get set. Would need to test.

peppy replied:
Added minimal error handling and deadlock avoidance.

bdach commented on osu.Game/Database/DetachedBeatmapStore.cs:
```diff
@@ -0,0 +1,154 @@
+        protected override void Dispose(bool isDisposing)
+        {
+            base.Dispose(isDisposing);
+            loaded.Set();
```
missing `Dispose()` call (`ManualResetEventSlim` is `IDisposable`)

bdach commented on osu.Game/Screens/Select/BeatmapCarousel.cs:
```diff
@@ -109,28 +110,38 @@ public partial class BeatmapCarousel : CompositeDrawable, IKeyBindingHandler<Glo
             {
-                loadedTestBeatmaps = true;
-                Schedule(() => loadBeatmapSets(value));
+                if (LoadState != LoadState.NotLoaded)
+                    throw new InvalidOperationException("If not using a realm source, beatmap sets must be set before load.");
```
If this sort of thing is gonna get thrown then maybe this entire property should be `internal` to begin with?

peppy replied:
Yeah fair. It's only used for testing. `internal` doesn't really protect us against much, but it's a visual hint I guess.

bdach commented on osu.Game/Screens/Select/BeatmapCarousel.cs:
```diff
@@ -345,9 +320,10 @@ private void processBeatmapChanges()
             try
             {
-                foreach (var set in setsRequiringRemoval) removeBeatmapSet(set);
+                // TODO: chekc whether we still need beatmap sets by ID
```
Should this todo be here?

peppy replied:
It shouldn't.

I checked on this. We kinda need this stuff locally to ensure we're handling the batch application correctly (see the schedule related comment above). If it wasn't for this we can simplify this class a lot more.

Probably a direction for the future.

bdach reviewed (commented):
just a read-through for now, haven't ran yet

peppy committed:
Remove TODO

peppy committed:
Mark `BeatmapSets` as `internal`

peppy committed:
Add disposal of `ManualResetEventSlim`

peppy committed:
Add very basic error handling for failed beatmap detach

peppy committed:
Fix beatmap not being detached on hide/unhide

The explicit detach call was removed from `updateBeatmapSet`, causing
this to occur. We could optionally add it back (it will be a noop in all
cases though).

bdach commented:
I guess the one thing about this diff is that it offloads the bulk of the load onto RAM, which can get quite sizeable:

![1725020717](https://github.com/user-attachments/assets/6e60da0b-0f1e-4bff-8758-6cdef16d7d70)

Granted, this is on a 234MB realm db, but 200MB is still nothing to scoff at when it comes to user machines. It's not *major*, but it's not spare change either.

That said it doesn't look like this is _new_, either, so I'm not sure cumulatively memory usage will go up, because this is how things look on `master` already:

![1725020974](https://github.com/user-attachments/assets/d784a768-c960-454e-9c4f-3e5c767dbce6)

bdach reviewed (approved):
Tested basic functionality I guess, nothing seems obviously broken.

Probably something to push out to users to see if it helps and hotfix-revert if it breaks.
